lib/populate.js

```diff
@@ -0,0 +1,341 @@
+const clone = require('clone')
+const async = require('async')
+const { concatArrays } = require('./utils')
+const { getRefModel, getSortedRefPaths } = require('./cbdocument.utils')
+const _ = require('lodash')
+const mpath = require('mpath')
+
+const DIGIT_REGEX = /^\d+$/
+
+/*!
+ * Checks whether an options parameter has valid populate options
+ * @param options
+ * @returns {*}
+ */
+function hasPopulate (options) {
+  if (!options) {
+    return false
+  }
+
+  if (typeof options.populate === 'undefined' || options.populate === null) {
+    return false
+  }
+
+  if (_.isString(options.populate)) {
+    return !_.isEmpty(options.populate)
+  }
+
+  if (_.isPlainObject(options.populate) && _.isString(options.populate.path)) {
+    return !_.isEmpty(options.populate.path)
+  }
+
+  if (_.isArray(options.populate)) {
+    const t = _.compact(options.populate)
+    return !_.isEmpty(t)
+  }
+
+  if (_.isBoolean(options.populate)) {
+    return options.populate
+  }
+
+  return false
+}
+
+function getArrayMap (path, ids, models, RefModel) {
+  let idArrayMap = {}
+  _.each(models, m => {
+    const p = m.get(path, true)
+    if (_.isString(p)) {
+      idArrayMap[m.getDocumentKeyValue(true)] = RefModel.getDocumentKeyValue(p)
+    }
+  })
+  return idArrayMap
+}
+
+/*!
+ * Populate everything
+ * @param refs
+ * @param options
+ * @param memo
+ * @param missing
+ * @param fn
+ */
+exports.populateAll = function populateAll (refs, options, memo, missing, fn) {
+  populateAllArray.call(this, [this], refs, options, memo, missing, (err, docs, missing) => {
+    return fn(err, docs ? docs[0] : undefined, missing)
+  })
+}
+
+function getIdsAndArrayMap (path, models, RefModel) {
+  let ids = _.map(models, path)
```

Comment:
Maybe it would make sense to chain all these lodash functions up for a lazy loading execution? 

```
const ids = _(models)
  .map(path)
  .compact()

//...

if (ids.size() === 0) {}

//...

ids.filter(_.isString)

//...

return {
  ids: ids.value()
}
```